src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -361,6 +361,124 @@ SGMatrix<T> element_prod(SGMatrix<T>& a, SGMatrix<T>& b)
 	return result;
 }
 
+/** Performs the operation C = A * B where "*" denotes matrix multiplication.
+ *
+ * This version returns the result in-place.
+ * User should pass an appropriately allocate memory matrix
+ *
+ * @param A First matrix
+ * @param B Second matrix
+ * @param result Result matrix
+ * @param transpose_A whether to transpose matrix A
+ * @param transpose_B whether to transpose matrix B
+ */
+template <typename T>
+void matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B, SGMatrix<T>& result,
+	bool transpose_A=false, bool transpose_B=false)
+{
+	REQUIRE(!(result.on_gpu()^A.on_gpu()),
+			"Cannot operate with matrix result on_gpu (%d) and \
+			 matrix A on_gpu (%d).\n", result.on_gpu(), A.on_gpu());
+	REQUIRE(!(result.on_gpu()^B.on_gpu()),
+			"Cannot operate with matrix result on_gpu (%d) and \
+			 matrix B on_gpu (%d).\n", result.on_gpu(), B.on_gpu());
+
+	if (transpose_A)
+	{
+		REQUIRE(A.num_cols == result.num_rows, "Number of columns for A (%d) and "
+				"number of rows for result (%d) should be equal!\n", A.num_cols, result.num_rows);
+		if (transpose_B)
+		{
+			REQUIRE(A.num_rows == B.num_cols, "Number of rows for A (%d) and "
+					"number of columns for B (%d) should be equal!\n", A.num_rows, B.num_cols);
+			REQUIRE(B.num_rows == result.num_cols, "Number of rows for B (%d) and "
+					"number of columns for result (%d) should be equal!\n",
+					B.num_rows, result.num_cols);
+		}
+		else
+		{
+			REQUIRE(A.num_rows == B.num_rows, "Number of rows for A (%d) and "
+					"number of rows for B (%d) should be equal!\n", A.num_rows, B.num_rows);
+			REQUIRE(B.num_cols == result.num_cols, "Number of columns for B (%d) and "
+					"number of columns for result (%d) should be equal!\n",
+					B.num_cols, result.num_cols);
+		}
+	}
+	else
+	{
+		REQUIRE(A.num_rows == result.num_rows, "Number of rows for A (%d) and "
+				"number of rows for result (%d) should be equal!\n", A.num_rows, result.num_rows);
+		if (transpose_B)
+		{
+			REQUIRE(A.num_cols == B.num_cols, "Number of columns for A (%d) and "
+					"number of columns for B (%d) should be equal!\n", A.num_cols, B.num_cols);
+			REQUIRE(B.num_rows == result.num_cols, "Number of rows for B (%d) and "
+					"number of columns for result (%d) should be equal!\n",
+					B.num_rows, result.num_cols);
+		}
+		else
+		{
+			REQUIRE(A.num_cols == B.num_rows, "Number of columns for A (%d) and "
+					"number of rows for B (%d) should be equal!\n", A.num_cols, B.num_rows);
+			REQUIRE(B.num_cols == result.num_cols, "Number of columns for B (%d) and "
+					"number of columns for result (%d) should be equal!\n",
+					B.num_cols, result.num_cols);
+		}
```

Comment:
I guess you could simplify via just saying that the dimensions don't match and then printing the given ones plus the transpose information, then you just need one message. But this verbose version has advantages from a user perspective as well..... leaving it to you